### PR TITLE
Added ability to send .zip files

### PR DESCRIPTION
**Pull request by Stwend**
I have added the function /unzip with which the user can send zipped files (both .zip and multi-volume .zip.001, zip.002, etc files) to the Discord bot. Only gcode files are extracted from the zip, after extraction the zip files are deleted and the gcode file is available for printing as usual.

By using zipped files it is possible to circumvent Discord's 8MB upload limit. The function allows for up to 99 multi-volume zips which can be 8MB each, enabling users to upload huge .gcode files.

How to use:

Case A - single-volume zip:
Call /unzip filename.zip

Case B - multi-volume zip:
Call /unzip filename.zip.001
The function will automatically search for the rest of the multi-volume.

If there are multiple gcode files in a zip, the function will extract them all.

! Make sure the Discord bot has accepted all of your files before calling /unzip !

### CODE REVIEW

File: octoprint_discordremote/__init__.py [hunk 1]
```diff
@@ -36,6 +36,17 @@ class DiscordRemotePlugin(octoprint.plugin.EventHandlerPlugin,
                           octoprint.plugin.TemplatePlugin,
                           octoprint.plugin.ProgressPlugin,
                           octoprint.plugin.SimpleApiPlugin):
+  
+    @property
```
Comment (cameroncros) [inline octoprint_discordremote/__init__.py:40]: What is the purpose of these changes?

File: octoprint_discordremote/command.py [hunk 2]
```diff
@@ -368,6 +371,87 @@ def download_file(self, filename, url, user):
         return None, success_embed(author=self.plugin.get_printer_name(),
                                    title='File Received',
                                    description=filename)
+    
+    
+    def unzip(self, params):
```
Comment (cameroncros) [inline octoprint_discordremote/command.py:376]: I havent manually tested yet, but could you please add a unit test for this?
See: https://github.com/cameroncros/OctoPrint-DiscordRemote/blob/master/unittests/test_command.py#L422 for an example, but a unit test that at least covers the successful unzip, and ideally covers all the return paths (if possible to test). I know writing unit tests sucks, but without them, it is hard to ensure that this code never breaks

**Review by cameroncros — CHANGES_REQUESTED** [review]
Overall, looks fantastic, and I would love to get this in. Can you please add a unit test for the function you have added, and explain the meaning of the __init__.py changes.

File: octoprint_discordremote/__init__.py (hunk 1 above)
Comment (Stwend) [reply octoprint_discordremote/__init__.py:40]: That's a leftover from when I stored the file type list as a variable instead of a function and needed a getter, I'll remove it.

File: octoprint_discordremote/command.py (hunk 2 above)
Comment (Stwend) [reply octoprint_discordremote/command.py:376]: Will do for sure, I'll have to add some dummy zips for the unit tests.

File: octoprint_discordremote/__init__.py (hunk 1 above)
Comment (Stwend) [reply octoprint_discordremote/__init__.py:40]: Ha, it just occurred to me that you might be talking about the entire bunch of lines changed, not just the @property decorator.
I had to add all of the zip filestpyes to Octoprint's list of allowed files or else Octoprint would instantly delete all of them upon receiving them:

allowed(self) is a function that provides a string containing a list of all of the zip filestpyes (.zip, .zip.001, .zip.002 etc)
get_extension_tree(self, *args, **kwargs) then adds that list to Octoprint's file types list.

I guess this could be written in a more elegant way, either by just building the list inside get_extension_tree (if the function is called only once) or building the list inside __init__ and storing it as a variable (if get_extension_tree is called multiple times).

File: octoprint_discordremote/command.py (hunk 2 above)
Comment (cameroncros) [reply octoprint_discordremote/command.py:376]: there is existing code for generating a zip file, you may want to steal that instead of committing large test files

File: octoprint_discordremote/__init__.py (hunk 1 above)
Comment (cameroncros) [reply octoprint_discordremote/__init__.py:40]: Cool, that makes a lot of sense, mayube chuck a comment in and its all good

Comment (Stwend) [reply octoprint_discordremote/__init__.py]: With Commit 3d1e4b9 I have added a comment and simplified the function a bit - it now doesn't build a string and turns it into a list, it builds the light right away.

File: octoprint_discordremote/command.py (hunk 2 above)
Comment (Stwend) [reply octoprint_discordremote/command.py:376]: With Commit I have added a unit test with checks for the following cases:
  Case 1: Requested zip file doesn't exist (unzip failure)
  Case 2: Requested file exists but is not a zip file (unzip failure)
  Case 3: Requested file is a zip file (unzip success)
          Also checks if the function only extracts .gcode files and ignores other file types
  Case 4: Requested file is a multi-volume zip file (unzip success)
  Case 5: Requested file is a multi-volume zip with missing volumes (unzip failure)

**Comment by cameroncros** [comment]
This all looks great. I will test and merge it. As a possible future suggestion, maybe when a zip file is uploaded, you could automatically try the unzip function, and quietly fail if not all files are ready. That way you can upload all parts of a zip, and when is ready, the GCODE files will be ready to be printed?

**Comment by Stwend** [comment]
> 
> 
> This all looks great. I will test and merge it. As a possible future suggestion, maybe when a zip file is uploaded, you could automatically try the unzip function, and quietly fail if not all files are ready. That way you can upload all parts of a zip, and when is ready, the GCODE files will be ready to be printed?

Thanks!
I've thought about auto-unzipping, too, my 2 cents:
It might be a little tricky to detect when multi-volume zips have finished uploading. The first file should have some info in its header about the overall size, though, and the multi-volume assembly code might lend itself to some easy "is-the-zip-working-yet testing". 
Alternatively, I could add a code snippet to the file downloader and make it judge the progress by adding up file sizes of received zip.xxx files until the file size in the zip.001 header is matched.
Rather than silently failing, I'd do a progress report, something along the lines of "5 of 8 files received", or "5 of ??? files received" if the plugin doesn't know the final file count yet.

It does sound like a good thing overall, I'll open a ticket, but I don't know when I'll be able to try it out :)

**Comment by cameroncros** [comment]
I figured that it could be detected just by trying to unzip, and catching the failure. But, I don't really know. Like it said, cool feature, but in no way is it essential or urgent.
